File: solutions/fsharp/two-bucket/1/TwoBucket.fs
// Dedicated to Shree DR.MDD
module TwoBucket

type Bucket = | One = 1 | Two = 2
type Outcome = { Moves:int; GoalBucket:Bucket; OtherBucket:int }

let private pour (x,y) (_,capY) = 
    let newY = if y + x > capY then capY else y + x
    let newX = if y + x <= capY then 0 else x - (capY - y)
    (newX, newY)

let private pourBack (x,y) (capX,_) = 
    let (resY,resX) = pour (y,x) (0,capX)
    (resX, resY)

let rec private explore path (cap1,cap2) goal forbidden current : Outcome option =
    let updatedPath = path @ [current]
    match current with
    | (x,y) when x = goal -> Some { Moves = updatedPath.Length; GoalBucket = Bucket.One; OtherBucket = y }
    | (x,y) when y = goal -> Some { Moves = updatedPath.Length; GoalBucket = Bucket.Two; OtherBucket = x }
    | (x,y) ->
        [(x,0);(0,y);(cap1,y);(x,cap2);pour (x,y) (cap1,cap2);pourBack (x,y) (cap1,cap2)]
        |> List.filter (fun state -> not (List.contains state updatedPath) && state <> forbidden)
        |> List.map (explore updatedPath (cap1,cap2) goal forbidden)
        |> List.choose id
        |> List.sortBy (fun o -> o.Moves)
        |> List.tryHead

let measure sizeA sizeB goal (start:Bucket) : Outcome =
    let initial = if start = Bucket.One then (sizeA,0) else (0,sizeB)
    let forbidden = if start = Bucket.One then (0,sizeB) else (sizeA,0)
    explore [] (sizeA,sizeB) goal forbidden initial
    |> function
    | Some result -> result
    | None -> failwith "Unsolvable"

File: solutions/julia/two-bucket/1/two-bucket.jl
# Dedicated to Shree DR.MDD

function twobucket(b1, b2, target, start)
    if target % gcd(b1, b2) != 0 || target > max(b1, b2)
        throw(DomainError("Impossible"))
    end

    moves = 1
    answer = nothing
    visited = Set([(0, 0), (b1, 0), (0, b2)])
    state = start == 1 ? Set([(b1, 0)]) : Set([(0, b2)])

    for (x, y) in state
        if x == target || y == target
            answer = (x, y)
        end
    end

    while answer === nothing
        prev = state
        state = Set()
        moves += 1
        for (x, y) in prev
            nxt = Set([(x, 0), (0, y), (x, b2), (b1, y)])

            t12 = min(x, b2 - y)
            push!(nxt, (x - t12, y + t12))

            t21 = min(y, b1 - x)
            push!(nxt, (x + t21, y - t21))

            for p in setdiff(nxt, visited)
                push!(state, p)
                push!(visited, p)
            end
        end

        for (x, y) in state
            if x == target || y == target
                answer = (x, y)
                break
            end
        end
    end

    return answer[1] == target ? (moves, 1, answer[2]) : (moves, 2, answer[1])
end
